src/server/tx_base.h

```diff
@@ -34,28 +35,50 @@ struct KeyLockArgs {
 
 // Describes key indices.
 struct KeyIndex {
-  unsigned start;
-  unsigned end;   // does not include this index (open limit).
-  unsigned step;  // 1 for commands like mget. 2 for commands like mset.
+  KeyIndex(unsigned start = 0, unsigned end = 0, unsigned step = 1,
+           std::optional<unsigned> bonus = std::nullopt)
+      : start(start), end(end), step(step), bonus(bonus) {
+  }
+
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = unsigned;
+  using difference_type = std::ptrdiff_t;
+  using pointer = value_type;
+  using reference = value_type;
 
-  // if index is non-zero then adds another key index (usually 0).
-  // relevant for for commands like ZUNIONSTORE/ZINTERSTORE for destination key.
-  std::optional<uint16_t> bonus{};
+  unsigned operator*() {
+    if (bonus)
+      return *bonus;
+    return start;
+  }
 
-  KeyIndex(unsigned s = 0, unsigned e = 0, unsigned step = 0) : start(s), end(e), step(step) {
+  KeyIndex& operator++() {
+    if (bonus)
```

Comment:
it's a bit dangerous and counter intuitive

Reply:
Then suggest a more elegant way 🙂 I thinks it's still better than writing `if (bonus)` in high level logic